pkg/apis/pipeline/v1/param_types.go

```diff
@@ -147,6 +147,15 @@ func (ps Params) extractParamMapArrVals() map[string][]string {
 	return paramsMap
 }
 
+// isArrayReference returns true if the parameter
+// is a reference to an array
+func (p Param) isArrayReference() bool {
+	if p.Value.Type != ParamTypeString {
+		return false
+	}
+	return strings.Contains(p.Value.StringVal, "[*]")
```

Comment:
don't think this is a valid check that it's an array reference, it would even consider object params as array references 🙃 

```yaml
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: pipeline-test
spec:
  params:
    - name: gitrepo
      properties:
        url: {}
        commit: {}
  tasks:
    - name: task1
      matrix:
        - name: arg
          value: $(params.gitrepo[*])
      taskRef:
        name: write-param-to-result
```

maybe we need to check the type of the referenced param as well 🤔 

or move the validation to after the parameter replacements have been applied, and check there that it's actually an array

Reply:
I see. I moved the logic for checking param types in a matrix to after applying parameters. @jerop